text/0015-interest-management.md

```diff
@@ -0,0 +1,258 @@
+
+## Integration Points
+Apart from the spawn / de-spawn connections mentioned above, the `InterestManager` integrates with the rest of the library in 2 places:
+
+* In `NetworkBehaviour::NetworkBehaviourUpdate()`. In the loop over clients the `InterestManager` is queried resulting in a list of objects.  Then this list is iterated over as the library prepares to transmit variable updates to the client.  This is basically a filter on top of the existing mechanism which updates all variables unconditionally
+* When RPCs are sent from the server, the library queries the `InterestManager` to get a list of objects that a given client can see, so that it can then determine if the object associated with the RPC is relevant to the client. 
+
+
+## Sample 1: Creating a Radius-based Node 
+
+Let's walk through a sample.  We will establish a node, then a kernel, then associate the kernel with the node and the node with a prefab.  For context while we're here, let's just look at an abbreviated version of the code for `InterestNodeStatic`:
+
+```
+public class InterestNodeStatic : InterestNode
+{
+    public List<InterestKernel> InterestKernels = new List<InterestKernel>();
+
+    // these are the objects under my purview
+    protected HashSet<NetworkObject> ManagedObjects;
+
+    public override void AddObject(in NetworkObject obj)
+    {
+        ManagedObjects.Add(obj);
+    }
+
+    public override void RemoveObject(in NetworkObject obj)
+    {
+        ManagedObjects.Remove(obj);
+    }
+
+    public override void QueryFor(in NetworkClient client, HashSet<NetworkObject> results)
+    {
+        if (InterestKernels.Count > 0)
+        {
+            foreach (var obj in ManagedObjects)
+            {
+                foreach (var ik in InterestKernels)
+                {
+                    ik.QueryFor(client, obj, results);
+                }
+            }
+        }
+        else
+        {
+            results.UnionWith(ManagedObjects);
+        }
+    }
+}
+```
+
+As we described earlier, this node maintains its own storage for the objects it tracks (`ManagedObjects`).  When it is told one of its associated objects is added, it simply adds it to its storage and vice versa when an object is going to be deleted.  At query time, if we have no kernels we simply merge the results passed in from the other clients with all the objects we manage and do no other computation.  Otherwise we invoke our kernel(s) and merge in those results.
+
+A default instance of this node, by the way, is instantiated in the `NetworkingManager` and is where objects go that have no associated `InterestNodes`.
+
+Now let's author the `RadiusInterestKernel`.  Actually, this file is included but let's look at its implementation:
+
+```
+public class RadiusInterestKernel : InterestKernel
```

Comment:
I'm trying to understand how `InterestKernels` would be used in a real world example. A lot of common approaches to AOI (spatial hashing, octrees, radix sort etc.) are stateful and calculate their state in a single pass over all objects, so the QueryFor interface of the `InterestKernel` won't really work for them. So from what I understand `InterestNode` would have to be used instead of `InterestKernel`, is that correct?

In the description about InterestStaticNode above it is mentioned that "for a given prefab we want to consider it relevant if it is within a given radius and on the same team as the querying connection's object" would be a use case. 
But from my understand this kernel approach is an `OR` and not an `AND` operation. So instead of `all objects within a given radius AND on the same team` we query for `all objects within a given radius OR in the same team`.

Reply:
Ok, good insight.  It is true, you would not likely do a domain decomposition by just standing up an `InterestKernel` and sticking it on a `InterestNodeStatic`.  However, like in the odds/evens example, what I would hope you would do is put the 'dispatch to the appropriate grid node' in a custom `InterestNode`.  I mean, that has to be custom.  But then, that node would then dispatch the correct grid node which (I think) would be `InterestNodeStatic` instances.  And if you wanted to do processing beyond just 'return all the things in this grid cell', then you'd attach an `InterestKernel` to them.   If not, just like the odds/evens example, the grid cell has no interest kernel and just returns everything.

Reply:
On your second comment, thank you, good point, I'll correct it, it is indeed an `OR` relationship and I meant it to be (e.g. I want to see my team members regardless if they're close by).  

But the next question I have for you & the internet, and @TwoTenPvP raised this to me in a 1:1, do we think it's important to let you define `AND` relationships?